### PR TITLE
Change how HTTP/2 frames are parsed and generated

### PR DESCRIPTION
**Pull request by Tratcher**
#2858 I was able to reduce the request buffer from 16kb to 15 bytes. The response buffer is still needed for compressing headers.

This helped eliminate several data copies for request headers, request and response data, and settings frames.

Converting Http2Frame to a POCO helped eliminate a lot of redundant de/serialization of properties that were accessed multiple times.

Measurements needed.

### CODE REVIEW

**Comment by Tratcher** [comment]
@geoffkizer FYI, are you using the Http2Frame type at all?

**Comment by geoffkizer** [comment]
Not currently, but thanks for the heads up

**Comment by Tratcher** [comment]
I had @sebastienros run this in the lab and he saw no performance change. That's not unexpected since this should only improve per connection memory and the standard perf tests focus on stream density.

File: src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs [hunk 1]
```diff
@@ -3,40 +3,245 @@
...
+
+        private const int LengthOffset = 0;
+        private const int TypeOffset = 3;
+        private const int FlagsOffset = 4;
+        private const int StreamIdOffset = 5;
+
+        public const int SettingSize = 6; // 2 bytes for the id, 4 bytes for the value.
+
+        // No frame type needs more than 8 bytes of additional fields. Payload data beyond the known fields is not stored in this buffer.
+        private readonly byte[] _extendedHeader = new byte[HeaderLength + 8];
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Why not make this a static field instead? That would allow us to keep this class static and I believe static classes have a small performance benefit. Did a quick test on a local branch and it seems to work.

Comment (JunTaoLuo) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Nevermind, we need a buffer per connection.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -606,7 +601,7 @@ private Task ProcessRstStreamFrameAsync()
             return Task.CompletedTask;
         }
 
-        private async Task ProcessSettingsFrameAsync()
+        private Task ProcessSettingsFrameAsync(ReadOnlySequence<byte> payload)
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: part of this was done in https://github.com/aspnet/KestrelHttpServer/commit/de5ccb5c78db1cfd13b93579738be6d80e0ecb08

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs [hunk 3]
```diff
@@ -109,14 +113,34 @@ public Task Write100ContinueAsync(int streamId)
...
-                _continueBytes.CopyTo(_outgoingFrame.HeadersPayload);
-
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                _outputWriter.Write(_continueBytes);
+                return _flusher.FlushAsync();
             }
         }
 
+        // TODO: Why doesn't this method flush? Is that because we always write a data frame afterwards (even if it's empty)?
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: I would assume so, but is there a penalty to flushing early?

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs [hunk 4]
```diff
@@ -182,45 +206,54 @@ public Task WriteDataAsync(int streamId, StreamOutputFlowControl flowControl, Re
...
+                foreach (var buffer in currentData)
                 {
-                    current.Span.CopyTo(payload);
-                    payload = payload.Slice(current.Length);
-                    unwrittenPayloadLength += current.Length;
+                    _outputWriter.Write(buffer.Span);
                 }
+
+                // Plus padding
+
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: nit: extra line

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs [hunk 5]
```diff
@@ -271,71 +304,194 @@ private async Task WriteDataAsyncAwaited(int streamId, StreamOutputFlowControl f
...
             {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
                 _outgoingFrame.PrepareRstStream(streamId, errorCode);
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                var buffer = _outputWriter.GetSpan(4);
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: These three lines writing a uint/31-bit int/short could be extracted to extension methods.

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs [hunk 6]
```diff
@@ -271,71 +304,194 @@ private async Task WriteDataAsyncAwaited(int streamId, StreamOutputFlowControl f
...
+                {
+                    return Task.CompletedTask;
+                }
+
+                _outgoingFrame.PrepareSettings(Http2SettingsFrameFlags.NONE);
+                var settingsSize = settings.Count * Http2FrameReader.SettingSize;
+                _outgoingFrame.PayloadLength = settingsSize;
+                WriteHeaderUnsynchronized();
+
+                var buffer = _outputWriter.GetSpan(settingsSize).Slice(0, settingsSize); // GetSpan isn't precise
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: How so? I'm wondering if it affects other places where GetSpan is called, especially on irregular numbers such as Http2FrameReader.HeaderLength

File: test/Kestrel.InMemory.FunctionalTests/Http2/Http2TestBase.cs [hunk 7]
```diff
@@ -345,90 +351,158 @@ protected async Task StartStreamAsync(int streamId, IEnumerable<KeyValuePair<str
...
 
-            var frame = new Http2Frame(Http2Limits.MinAllowedMaxFrameSize);
+            var frame = new Http2Frame();
             frame.PrepareHeaders(Http2HeadersFrameFlags.END_HEADERS | Http2HeadersFrameFlags.PADDED | Http2HeadersFrameFlags.PRIORITY, streamId);
             frame.HeadersPadLength = padLength;
             frame.HeadersPriorityWeight = priority;
             frame.HeadersStreamDependency = streamDependency;
 
-            _hpackEncoder.BeginEncode(headers, frame.HeadersPayload, out var length);
+            var extendedHeaderLength = 6; // stream dependency + weight
```
Comment (JunTaoLuo) [inline test/Kestrel.InMemory.FunctionalTests/Http2/Http2TestBase.cs]: nit: padding + ..

File: test/Kestrel.InMemory.FunctionalTests/Http2/Http2TestBase.cs [hunk 8]
```diff
@@ -345,90 +351,158 @@ protected async Task StartStreamAsync(int streamId, IEnumerable<KeyValuePair<str
...
+            var buffer = _headerEncodingBuffer.AsSpan();
+            var extendedHeader = buffer.Slice(0, extendedHeaderLength);
+            extendedHeader[0] = padLength;
+            var payload = buffer.Slice(extendedHeaderLength, buffer.Length - padLength - extendedHeaderLength);
+
+            _hpackEncoder.BeginEncode(headers, payload, out var length);
+            var padding = buffer.Slice(extendedHeaderLength + length, padLength);
+            padding.Fill(0);
 
             frame.PayloadLength = 1 + length + padLength;
```
Comment (JunTaoLuo) [inline test/Kestrel.InMemory.FunctionalTests/Http2/Http2TestBase.cs]: nit: 1 => extendedHeaderLength

File: src/Kestrel.Core/Internal/Http2/HPack/HPackDecoder.cs [hunk 9]
```diff
@@ -108,11 +109,15 @@ internal HPackDecoder(int maxDynamicTableSize, DynamicTable dynamicTable)
             _dynamicTable = dynamicTable;
         }
 
-        public void Decode(Span<byte> data, bool endHeaders, IHttpHeadersHandler handler)
+        public void Decode(ReadOnlySequence<byte> data, bool endHeaders, IHttpHeadersHandler handler)
         {
-            for (var i = 0; i < data.Length; i++)
+            foreach (var segment in data)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http2/HPack/HPackDecoder.cs]: This is going to be inefficient, but I'm not sure it's worth it to fix this until we remove OnByte

File: src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs [hunk 10]
```diff
@@ -3,40 +3,245 @@
...
+                case Http2FrameType.CONTINUATION: // None
+                case Http2FrameType.PUSH_PROMISE: // Not implemented frames are ignored at this phase
+                default:
+                    return 0;
+            }
+        }
+
+        public IList<Http2PeerSetting> ReadSettings(ReadOnlySequence<byte> payload)
+        {
+            var data = payload.ToArray().AsSpan();
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: This is super inefficient. There's a GetSpan helper in Kestrel that only allocates if it needs to. Use that.

File: src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs [hunk 11]
```diff
@@ -3,40 +3,245 @@
...
-            if (readableBuffer.Length < Http2Frame.HeaderLength)
+            if (readableBuffer.Length < HeaderLength)
             {
                 return false;
             }
 
-            var headerSlice = readableBuffer.Slice(0, Http2Frame.HeaderLength);
-            headerSlice.CopyTo(frame.Raw);
+            var headerSlice = readableBuffer.Slice(0, HeaderLength);
+            headerSlice.CopyTo(_extendedHeader);
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Why do you copy at all?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Because parsing out of a `ReadOnlySequence<byte>` looked more complex than these 15 bytes were worth.

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: You don't need to copy though, you can just do what we do elsewhere.  Use the [ToSpan](https://github.com/aspnet/KestrelHttpServer/blob/d3d7c5519810ad97997f777c8a3683a9adbbe1e1/src/Kestrel.Core/Internal/Http/PipelineExtensions.cs#L20) extension method and 80% of the time it won't allocate. 😄

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: That's a question of odds then. One allocation per connection, or one allocation every time it ToSpan fails. If it fails more than once per connection then you're net negative.

Also we have to do it in two segments since the extended headers are variable length.

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs (hunk 3 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: Flushing early creates more TLS & TCP packets.

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs (hunk 5 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: If there were duplicates it would be worth extracting, but 
every usage of WriteUInt32BigEndian ends up being a little different.

File: src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs (hunk 11 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Does this logic run once per connection?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: No, it's on every frame.

File: src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs (hunk 6 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs]: GetSpan gives you the current buffer if your hint size will fit in it, otherwise it gives you the next buffer. It doesn't slice it down to the size your asked for though. It's also limited to 4kb.

File: src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs (hunk 11 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs]: Then it's worth changing.

**Comment by Tratcher** [comment]
Rebased and updated

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 12]
```diff
@@ -639,14 +633,15 @@ private Task ProcessSettingsFrameAsync()
...
 
                 // Ack before we update the windows, they could send data immediately.
-                var ackTask =  _frameWriter.WriteSettingsAckAsync();
+                var ackTask = _frameWriter.WriteSettingsAckAsync();
 
                 if (_clientSettings.MaxFrameSize != previousMaxFrameSize)
                 {
-                    _frameWriter.UpdateMaxFrameSize(_clientSettings.MaxFrameSize);
+                    // Don't let the client choose an arbitrarily large size, this will be used for response buffers.
+                    _frameWriter.UpdateMaxFrameSize(Math.Min(_clientSettings.MaxFrameSize, _serverSettings.MaxFrameSize));
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Can we add a new test to ensure the minimum of the two values is always picked?

File: src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs [hunk 13]
```diff
@@ -24,20 +24,29 @@ public bool BeginEncode(int statusCode, IEnumerable<KeyValuePair<string, string>
...
         {
             length = 0;
 
             do
             {
                 if (!EncodeHeader(_enumerator.Current.Key, _enumerator.Current.Value, buffer.Slice(length), out var headerLength))
                 {
+                    if (length == 0 && throwIfNoneEncoded)
+                    {
+                        throw new ArgumentOutOfRangeException("The given buffer was too small to encode any headers.");
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: Yes it does surface to the application on first write. I added it because I hung multiple tests while refactoring the max frame size handling. Without it this code spins in a loop.

I'm not as concerned about the application running into this. The client isn't allowed to reduce the max frame size bellow 16kb, and if your individual headers are over 16kb then you have other issues. I will change the type to InvalidOperationException.

[the remark this answers is not in the document]
Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: How is it logged if the response doesn't contain a body?

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: It's logged as a general request error. I've added a little more error handling to abort the connection since the header compression table would now be considered corrupt. There's not a good way to run the normal Http2Connection cleanup code from here. I'm not concerned with making this code path more graceful at the moment, we'll see if users really run into this scenario.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 14]
```diff
@@ -422,12 +426,6 @@ private Task ProcessDataFrameAsync()
                 throw new Http2ConnectionErrorException(CoreStrings.FormatHttp2ErrorStreamIdZero(_incomingFrame.Type), Http2ErrorCode.PROTOCOL_ERROR);
             }
 
-            // The Padding field is missing
-            if (_incomingFrame.DataPayloadOffset > _incomingFrame.PayloadLength)
-            {
-                throw new Http2ConnectionErrorException(CoreStrings.Http2FrameMissingFields, Http2ErrorCode.PROTOCOL_ERROR);
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: The frame reader checks it while it's reading those extra fields: https://github.com/aspnet/KestrelHttpServer/pull/2893/files#diff-feb5d4f5a5878a3a2366ddb77ef8d020R79

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 12 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Added

**Comment by Tratcher** [comment]
Rebased and updated.

File: src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: Filed https://github.com/aspnet/KestrelHttpServer/issues/2926 to address this by splitting header values acrross frames.

File: src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs [hunk 15]
```diff
@@ -24,20 +24,29 @@ public bool BeginEncode(int statusCode, IEnumerable<KeyValuePair<string, string>
...
         {
             length = 0;
 
             do
             {
                 if (!EncodeHeader(_enumerator.Current.Key, _enumerator.Current.Value, buffer.Slice(length), out var headerLength))
                 {
+                    if (length == 0 && throwIfNoneEncoded)
+                    {
+                        throw new HPackEncodingException(CoreStrings.HPackErrorNotEnoughBuffer);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: I know we don't plan for this to be permanent, but I would still like a test verifying this gets thrown from the first body write.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs]: Added

**Comment by davidfowl** [comment]
👍
